### PR TITLE
improvement(tree): `getJsonSchema` supports field schema input

### PR DESCRIPTION
**Pull request by Josmithr**
The API previously only supported `ImplicitAllowedTypes` input, but did not allow for `ImplicitFieldSchema` input. This meant that it could not be called via `getJsonSchema(node.schema)`.

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/api/getJsonSchema.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ const jsonSchemaCache = new WeakMap<TreeNodeSchema, JsonTreeSchema>();
  * 			"required": ["foo"],
  * 		},
  * 	},
- * 	"anyOf": [ { "$ref": "#/$defs/com.myapp.MyObject" } ],
+ * 	"$ref": "#/$defs/com.myapp.MyObject",
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/getJsonSchema.ts:56]: This example was out of date.

File: packages/dds/tree/src/simple-tree/api/simpleSchemaToJsonSchema.ts [hunk 2]
```diff
@@ -42,15 +42,18 @@ export function toJsonSchema(schema: SimpleTreeSchema): JsonTreeSchema {
...
+			? {
+					...(allowedTypes[0] ?? oob()),
+					$defs: definitions,
+				}
+			: {
+					$defs: definitions,
+					anyOf: allowedTypes,
+				};
+
+	return output;
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/simpleSchemaToJsonSchema.ts:56]:
```suggestion
	return allowedTypes.length === 1
			? {
					...(allowedTypes[0] ?? oob()),
					$defs: definitions,
				}
			: {
					$defs: definitions,
					anyOf: allowedTypes,
				};
```

File: packages/dds/tree/src/simple-tree/api/simpleSchemaToJsonSchema.ts [hunk 3]
```diff
@@ -43,14 +43,14 @@ export function toJsonSchema(schema: SimpleTreeSchema): JsonTreeSchema {
...
-				anyOf: allowedTypes,
-			};
+			? {
+					...(allowedTypes[0] ?? oob()),
+					$defs: definitions,
+				}
+			: {
+					$defs: definitions,
+					anyOf: allowedTypes,
+				};
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/api/simpleSchemaToJsonSchema.ts:53]:
```suggestion
		? {
				...(allowedTypes[0] ?? oob()),
				$defs: definitions,
			}
		: {
				$defs: definitions,
				anyOf: allowedTypes,
			};
```

File: packages/dds/tree/src/simple-tree/api/simpleSchema.ts [hunk 4]
```diff
@@ -125,19 +125,11 @@ export interface SimpleFieldSchema {
  *
  * @sealed
  */
-export interface SimpleTreeSchema {
+export interface SimpleTreeSchema extends SimpleFieldSchema {
```
Comment (alexvy86) [inline packages/dds/tree/src/simple-tree/api/simpleSchema.ts:128]: This gives `SimpleTreeSchema` the new `kind` property... Does it conceptually mean the kind for the root field for the tree? Related, can `SimpleTreeSchema` represent subtrees (and in that case, `kind` means the kind of the field in the parent node, which acts as the root field for the subtree)?

File: packages/dds/tree/src/simple-tree/api/simpleSchema.ts [hunk 5]
```diff
@@ -125,19 +125,11 @@ export interface SimpleFieldSchema {
...
 	/**
 	 * The complete set of node schema definitions recursively referenced by the tree's {@link SimpleTreeSchema.allowedTypes}.
 	 *
 	 * @remarks the keys are the schemas' {@link TreeNodeSchemaCore.identifier | identifiers}.
 	 */
 	readonly definitions: ReadonlyMap<string, SimpleNodeSchema>;
-
-	/**
-	 * The types allowed under the root of the tree.
```
Comment (alexvy86) [inline packages/dds/tree/src/simple-tree/api/simpleSchema.ts:137]: This doc seems more appropriate for SimpleTreeSchema than the version it now gets from SimpleFieldSchema. We can keep the redundant definition here to give it more contextually relevant docs, right? If so, maybe do the same think for `kind`.

File: packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts [hunk 6]
```diff
@@ -15,6 +15,37 @@ import { hydrate } from "./utils.js";
...
+
+		const expected: JsonTreeSchema = {
+			$defs: {
+				"com.fluidframework.leaf.string": {
+					type: "string",
+					_treeNodeSchemaKind: NodeKind.Leaf,
+				},
+			},
+			$ref: "#/$defs/com.fluidframework.leaf.string",
+		};
```
Comment (alexvy86) [inline packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts:34]: Tangent: this doesn't seem to indicate that the string is optional, correct? If it does, I don't understand how, so I was wondering for my own education. But also, on line 46 it seems like we're validating that assigning `null` isn't valid. Should we validate what happens when assigning `undefined`?

**Review by alexvy86 — COMMENTED** [review]
Left a few comments. I'm not familiar enough with this part of the codebase to know if the change is correct, but I do wonder if `SimpleTreeSchema` should look the same as `SimpleFieldSchema` but not extend it, for conceptual clarity.

File: packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts (hunk 6 above)
Comment (seanimam) [reply packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts:34]: Required fields are included at the end of a definition like so:

```json
{
  "$id": "https://example.com/address.schema.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "description": "An address similar to http://microformats.org/wiki/h-card",
  "type": "object",
  "properties": {
    "postOfficeBox": {
      "type": "string"
    },
    "extendedAddress": {
      "type": "string"
    },
    "streetAddress": {
      "type": "string"
    },
    "locality": {
      "type": "string"
    },
    "region": {
      "type": "string"
    },
    "postalCode": {
      "type": "string"
    },
    "countryName": {
      "type": "string"
    }
  },
  "required": [ "locality", "region", "countryName" ],
  "dependentRequired": {
    "postOfficeBox": [ "streetAddress" ],
    "extendedAddress": [ "streetAddress" ]
  }
}
```
So, I would imagine some kind of a test here would check for the existence of the 'required' field and its values right?

**Comment by Josmithr** [comment]
> Left a few comments. I'm not familiar enough with this part of the codebase to know if the change is correct, but I do wonder if `SimpleTreeSchema` should look the same as `SimpleFieldSchema` but not extend it, for conceptual clarity.

In theory, it could - any level of the JSON Schema tree may contain $defs. But in our case, we always populate all $defs at the root. And there are systems that require this as an invariant (including OpenAI for structured output). So making the $defs field root-specific made the most sense to me.

File: packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts (hunk 6 above)
Comment (Josmithr) [reply packages/dds/tree/src/test/simple-tree/getJsonSchema.spec.ts:34]: Our unit tests compare exact JSON output, so this is covered. But yes, JSON Schema validators check the `required` property for this. Items not enumerated in that property are implicitly optional.

File: packages/dds/tree/src/simple-tree/api/simpleSchema.ts (hunk 5 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/api/simpleSchema.ts:137]: Fair. I'll update.

File: packages/dds/tree/src/simple-tree/api/simpleSchema.ts (hunk 4 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/api/simpleSchema.ts:128]: > This gives SimpleTreeSchema the new kind property... Does it conceptually mean the kind for the root field for the tree?

Yes. This was a gap in the original implementation. The root field of a tree may be optional (before, this code assumed it was always required). Fundamentally, the root of the tree is a field, not a node. The code previously was structured around the root input being a node (which it then mapped to an implicitly required field). This change makes this representation more consistent with the SharedTree's underlying representation.

> Related, can SimpleTreeSchema represent subtrees (and in that case, kind means the kind of the field in the parent node, which acts as the root field for the subtree)?

Yeah, basically. Really, it makes the root less special, so it's like any other field location in the tree.

**Comment by Josmithr** [comment]
> I do wonder if `SimpleTreeSchema` should look the same as `SimpleFieldSchema` but not extend it, for conceptual clarity.

I considered that. But the root of the tree _is_ a field. It also has the global definitions set, but it is still fundamentally a field, so the relationship made sense to me.

File: packages/dds/tree/src/simple-tree/api/viewSchemaToSimpleSchema.ts [hunk 7]
```diff
@@ -26,18 +27,26 @@ import { NodeKind, type TreeNodeSchema } from "../core/index.js";
...
 	populateSchemaDefinitionsForField(normalizedSchema, definitions);
 
-	return {
+	const output: Mutable<SimpleTreeSchema> = {
+		kind: normalizedSchema.kind,
 		allowedTypes,
 		definitions,
 	};
+
+	// Don't include "description" property at all if it's not present on the input.
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/api/viewSchemaToSimpleSchema.ts:44]:
```suggestion
	// Include the "description" property only if it's present on the input.
```

**Review by noencke — APPROVED** [review]

**Pull request merged**